src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionAnnotatedTypeFactory.java

```diff
@@ -186,13 +242,128 @@ private void handleToMethodInvocation(
+        if (AnnotationUtils.areSameByName(
+            anno, "org.checkerframework.checker.dependencyinjection.qual.BindAnnotatedWith")) {
+          bindAnno = anno;
+          break;
+        }
+      }
+    }
+
+    // Handle cases where there is a call to annotatedWith
+    if (bindAnno != null) {
+      List<String> boundClassNames =
+          AnnotationUtils.getElementValueArray(bindAnno, bawValValueElement, String.class);
+
+      // TODO: Possibly create a wrapper class for value
+      List<String> annotatedNames =
+          AnnotationUtils.getElementValueArray(
+              bindAnno, bawAnnotatedWithValueElement, String.class);
+
+      boundClassNames.forEach(
+          boundClassName -> {
+            if (DependencyInjectionAnnotatedTypeFactory.knownBindings.containsKey(boundClassName)) {
+              DependencyInjectionAnnotatedTypeFactory.knownBindings.remove(boundClassName);
+              // Class that is being bound to - put in knownBindings
+              // <bound,boundTo>
+              TypeMirror boundToClassName = methodArgumentNode.getType();
+
+              KnownBindingsValue knowBindingValue =
+                  new KnownBindingsValue.KnownBindingsValueBuilder(boundToClassName.toString())
+                      .setAnnotationName(annotatedNames.get(0))
+                      .build();
+
+              DependencyInjectionAnnotatedTypeFactory.knownBindings.put(
+                  boundClassName, knowBindingValue);
+            }
+          });
+    }
+  }
+
+  /**
+   * Invoked when the {@code MethodInvocationNode} is a call to {@code
+   * com.google.inject.binder.LinkedBindingBuilder.toInstance}
+   *
+   * <p>This method will attempt to find the class that is being bound and add it to the {@code
+   * knownBindings} map alongside the class that is being bound to
+   *
+   * <p>If the receiver is a call to {@code
+   * com.google.inject.binder.AnnotatedBindingBuilder.annotatedWith}, a call to {@code
+   * handleAnnotatedWithMethod} will be made
+   *
+   * @param currentNode the current node in the block
+   * @param methodAccessNode the method access node
+   * @param methodArgumentNode the argument to the{@code to} method
+   */
+  private void handleToInstanceMethodInvocation(
+      Node currentNode, MethodAccessNode methodAccessNode, Node methodArgumentNode) {
+
+    Node receiver = methodAccessNode.getReceiver();
+
+    if (isAnnotatedWithMethod(receiver.getTree())) {
+      handleAnnotatedWithMethod(currentNode, methodAccessNode, methodArgumentNode);
+    } else if (isBindMethod(receiver.getTree())) {
+      CFValue value = this.getStoreBefore(currentNode).getValue(JavaExpression.fromNode(receiver));
+
+      AnnotationMirror bindAnno = null;
+      if (value != null && !value.getAnnotations().isEmpty()) {
+        for (AnnotationMirror anno : value.getAnnotations()) {
+          if (AnnotationUtils.areSameByName(
+              anno, "org.checkerframework.checker.dependencyinjection.qual.Bind")) {
```

Comment:
What if the receiver of `toInstance` has an `@BAW` annotation? I think this code needs to explicitly account for that possibility, too.

Reply:
I believe that it is taking care of this scenario. The only time the receiver of `toInstance` can have a `@BAW` annotation is if it's a call to `annotatedWith`. If the receiver is a call to `annotatedWith`, it's taken care of on line **329**. If it's just a call to `bind`, it's taken care of on line **330**.

Reply:
I see.

In that case, I don't think this code should be checking for whether the receiver is a call to `bind()` or `annotatedWith` at all: instead, it should just look at the type of the receiver expression in the store and then branch based on whether it's `@Bind`/`@BAW`. That will allow us to properly handle cases where the result of `bind()` or `annotatedWith()` is stored into a local, etc., which is definitely desirable. As written, your system will only be able to verify fluent chains of calls, but the underlying analysis should be more general than that.

Reply:
I think I understand what you're getting at. By checking whether the receiver of the current node we're on is a call to `bind` or `annotatedWith`, we're restricting the algorithm to only work when the calls to these methods are chained. So, the program wouldn't be able to handle cases such as:
```java
      AnnotatedBindingBuilder<Integer> b1 = bind(Integer.class);
      LinkedBindingBuilder<Integer> lb1 = b1.annotatedWith(Names.named("Baz"));
      lb1.toInstance(10);
```

Reply:
Exactly. The algorithm should be general enough to handle that example (maybe you should add it as a test case!).